### ISSUE TITLE
Memoize save_function

### ISSUE THREAD

**Issue by mrocklin** [cloudpipe/cloudpickle#45]
I often need to serialize many small objects containing many python functions. 

``` python
In [1]: def inc(x):
    return x + 1
   ...: 

In [2]: d = {i: (inc, i) for i in range(10000)}
```

Sometimes I do this all at once; this works great.

``` python
In [3]: from cloudpickle import dumps, loads

In [4]: %time len(dumps(d))
CPU times: user 118 ms, sys: 0 ns, total: 118 ms
Wall time: 117 ms
```

But sometimes I do this in several small batches, which is much slower.

``` python
In [5]: %time len([dumps(item) for item in d.items()])
CPU times: user 2.7 s, sys: 3.93 ms, total: 2.7 s
Wall time: 2.71 s
```

A quick profile shows that the majority of time is spent in `save_function`

```
In [7]: %prun -s cumtime len([dumps(item) for item in d.items()])
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    4.782    4.782 {built-in method exec}
        1    0.001    0.001    4.782    4.782 <string>:1(<module>)
        1    0.038    0.038    4.782    4.782 <string>:1(<listcomp>)
    10000    0.025    0.000    4.744    0.000 cloudpickle.py:598(dumps)
    10000    0.011    0.000    4.658    0.000 cloudpickle.py:104(dump)
    10000    0.030    0.000    4.646    0.000 pickle.py:401(dump)
450000/10000    0.894    0.000    4.597    0.000 pickle.py:460(save)
120000/10000    0.287    0.000    4.568    0.000 pickle.py:716(save_tuple)
50000/10000    0.115    0.000    4.296    0.000 cloudpickle.py:162(save_function)
    10000    0.053    0.000    4.254    0.000 cloudpickle.py:214(save_function_tuple)
    10000    0.020    0.000    2.834    0.000 cloudpickle.py:142(save_codeobject)
40000/10000    0.120    0.000    2.814    0.000 cloudpickle.py:470(save_reduce)
50000/40000    0.117    0.000    1.285    0.000 cloudpickle.py:318(save_global)
    20000    0.039    0.000    1.058    0.000 pickle.py:680(save_bytes)
   290000    0.519    0.000    1.044    0.000 pickle.py:416(memoize)
    40000    0.257    0.000    0.716    0.000 pickle.py:898(save_global)
    70000    0.147    0.000    0.479    0.000 pickle.py:698(save_str)
   800000    0.270    0.000    0.392    0.000 pickle.py:212(write)
```

And so I'm tempted to memoize `save_function` between dumps calls.  Presumably with some sort of LRU mechanism, keying by object identity.  This is unsafe if functions mutate in any way.  I've never run into such a situation but I'm unsure if it's done elsewhere.

On looking into cloudpickle more deeply, it appears that `Pickler` has a caching mechanism within it.  Does anyone have experience with these `memo` objects?  I would need to clear out non-function elements from the cache between calls.

I'm happy to do the work here if we are able to agree on a good solution.

**Comment by mrocklin** [comment]
With the following trivial patch:

```
 import operator
+from toolz import memoize
 import os
 import io
 import pickle
@@ -159,6 +160,7 @@ class CloudPickler(Pickler):
         self.save_reduce(types.CodeType, args, obj=obj)
     dispatch[types.CodeType] = save_codeobject
 
+    @memoize(key=lambda args, kwargs: args[1])
     def save_function(self, obj, name=None):
```

I get decent speedups

```python
In [1]: def inc(x):
    return x + 1
   ...: 

In [2]: from cloudpickle import CloudPickler, dumps, loads

In [3]: d = {i: (inc, i) for i in range(10000)}

In [4]: %time len([dumps(item) for item in d.items()])  # used to take 3s
CPU times: user 251 ms, sys: 0 ns, total: 251 ms
Wall time: 250 ms
Out[4]: 10000
```

**Comment by rgbkrk** [comment]
Matt, I more than trust in your abilities and presume you're aiming for the best here. I don't have experience with these `memo` objects and that's ok. I'm happy to learn along the way here. I'm curious about the effect across a typical set of functions, though we don't have any previous metrics or test suite to evaluate this. I think this is a start!

How does this effect memory usage? How large is the LRU cache in `memoize` from `toolz`?

In case that came out too wordy or late night mumbled, I welcome the patch.

**Comment by mrocklin** [comment]
Memoize just uses a straight dictionary.  I don't think it should be the final solution.  I just wanted something small to use.

Regarding memory use functions are *usually* pretty small.  Cloudpickle is memoizing them anyway per `dumps` call so there is decent precedent here.  The proposed change is to optionally memoize across `dumps` calls.

**Comment by ogrisel** [comment]
I assume that if you change the definition of a function, it's physical address is changed as well and the cache should be invalidated right?

```python
>>> def f(x): return x ** 2
>>> id(f)
4355475656
>>> def f(x): return x ** 3
>>> id(f)
4355477152
```

+1 for an optional kwarg for dumps for a more aggressive memoization that is turned off by default.

**Comment by ogrisel** [comment]
Or maybe it's possible to scope the memoizer's cache to a specific `Pickler` instance instead of sharing a state at the dumps function level?

**Comment by pitrou** [comment]
> Or maybe it's possible to scope the memoizer's cache to a specific Pickler instance instead of sharing a state at the dumps function level?

If you're reusing the same Pickier instance, it will naturally memoize objects, so an additional cache would be pointless.
```
>>> def inc(x): return x + 1
>>> 
>>> from cloudpickle import Pickler
>>> from io import BytesIO
>>> bio = BytesIO()
>>> p = Pickler(bio)
>>> p.dump(inc)
>>> p.dump(inc)
>>> bio.getvalue()
b'\x80\x03c__main__\ninc\nq\x00.\x80\x03h\x00.'
>>> import pickletools
>>> pickletools.dis(bio.getvalue())
    0: \x80 PROTO      3
    2: c    GLOBAL     '__main__ inc'
   16: q    BINPUT     0
   18: .    STOP
highest protocol among opcodes = 2
```

**Comment by pitrou** [comment]
As for the general approach: the pickle output can depend on specifics: not only the protocol number, but the current `memo` dict (since identical objects get referenced by their insertion number in the memo dict). Not to mentions variations in the function itself: the values of global and closure variables for example. It means that `save_function_tuple` is too broad a target.

**Comment by pitrou** [comment]
Ok, so I figure this issue can just be closed?
